Libraries/LibWeb/IndexedDB/IDBFactory.cpp

```diff
@@ -173,4 +177,55 @@ WebIDL::ExceptionOr<GC::Ref<IDBOpenDBRequest>> IDBFactory::delete_database(Strin
     return request;
 }
 
+// https://w3c.github.io/IndexedDB/#dom-idbfactory-databases
+GC::Ref<WebIDL::Promise> IDBFactory::databases()
+{
+    auto& realm = this->realm();
+
+    // 1. Let environment be this's relevant settings object.
+    auto& environment = HTML::relevant_settings_object(*this);
+
+    // 2. Let storageKey be the result of running obtain a storage key given environment.
+    //    If failure is returned, then return a promise rejected with a "SecurityError" DOMException
+    auto maybe_storage_key = StorageAPI::obtain_a_storage_key(environment);
+    if (!maybe_storage_key.has_value())
+        return WebIDL::create_rejected_promise_from_exception(realm, WebIDL::SecurityError::create(realm, "Failed to obtain a storage key"_string));
+
+    auto storage_key = maybe_storage_key.release_value();
+
+    // 3. Let p be a new promise.
+    auto p = WebIDL::create_promise(realm);
+
+    // 4. Run these steps in parallel:
+    Platform::EventLoopPlugin::the().deferred_invoke(GC::create_function(realm.heap(), [&realm, storage_key, p]() {
+        HTML::TemporaryExecutionContext context(realm, HTML::TemporaryExecutionContext::CallbacksEnabled::Yes);
+
+        // 1. Let databases be the set of databases in storageKey.
+        //    If this cannot be determined for any reason, then reject p with an appropriate error (e.g. an "UnknownError" DOMException) and terminate these steps.
+        auto databases = Database::for_key(storage_key);
+
+        // 2. Let result be a new list.
+        Vector<JS::Value> result;
```

Comment:
Hmm, I might be wrong in my understanding, but this does not seem safe.

This contains a GC allocated object stored on the heap which does not seem to be visited anywhere. So the JS::Values might have a GC run of them during the loop over databases.

Reply:
What I think would be safe is if we did:

```cpp
auto array = MUST(Array::create(realm, 0));
```

here, and in the:

```
3. For each db of databases:
```
loop, we do:

```
MUST(array->create_data_property_or_throw(n, elements[n]));
``` 

That way, `array` will be visited on the GC stack scan